src/common/perf_counters_key.cc

```diff
@@ -0,0 +1,223 @@
+#include <algorithm>
+#include <iterator>
+#include <numeric>
+#include "common/perf_counters_key.h"
+
+namespace ceph::perf_counters {
+namespace detail {
+
+// use a null character to delimit strings
+constexpr char delimiter = '\0';
+
+
+// write a delimited string to the output
+auto write(std::string_view str, std::output_iterator<char> auto out)
+{
+  out = std::copy(str.begin(), str.end(), out);
+  *(out++) = delimiter;
+  return out;
+}
+
+// return the encoded size of a label
+inline std::size_t label_size(const label_pair& l)
+{
+  return l.first.size() + sizeof(delimiter)
+      + l.second.size() + sizeof(delimiter);
```

Comment:
(replying because this came from my https://github.com/ceph/ceph/pull/47652)

> Taking sizeof of the delimiter is confusing

i'd argue that magic numbers like `+ 2` are potentially more confusing, and would deserve a comment to explain

you're right that `sizeof(delimiter)` will always be 1, but i still prefer to document that we're adding to the length *because* of the delimiters

Reply:
In one other case where there would be ` + 1` according to my suggestion, there already is a "add a delimiter" comment.  Here it could be `l.first.size() + 1 + l.second.size() + 1` for extra clarity.  This is all string-handling code so `+ 1` or `- 1` is almost never confusing IMO.  But it's just a nit and completely up to you --  feel free to ignore and resolve related conversations.